docs/system-events-and-plugin-system.md

```diff
@@ -0,0 +1,47 @@
+# System Events and Plugin System
+
+## System Events
+
+System Events are special type of events emitted by the Event Gateway instance internally. They are emitted on each stage
+of event processing flow starting from receiving event to function invocation end. Those events are:
+
+* `gateway.event.received` - the event is emitted when an event was received by Events API. Data fields:
+  * `event` - event payload
+  * `path` - Events API path
+  * `headers` - HTTP request headers
+* `gateway.function.invoking` - the event emitted before invoking a function. Data fields:
+  * `event` - event payload
+  * `functionId` - registered function ID
+* `gateway.function.invoked` - the event emitted after successful function invocation. Data fields:
+  * `event` - event payload
+  * `functionId` - registered function ID
+  * `result` - function response
+* `gateway.function.invocationFailed` - the event emitted after failed function invocation. Data fields:
+  * `event` - event payload
+  * `functionId` - registered function ID
+  * `error` - invocation error
+
+## Plugin System
+
+The Event Gateway is built with extensibility in mind. Built-in plugin system allows reacting on system events and
+manipulate how an event is processed through the Event Gateway.
+
+_Current implementation supports plugins written only in Golang. We plan to support other languages in the future._
+
+Plugin system is based on [go-plugin](https://github.com/hashicorp/go-plugin). A plugin needs to implement the following
+interface:
```

Comment:
To make it easier, I think we should do a post on "Writing a plugin for EG". That way we can show how to write a plugin and also drive some traffic.

Reply:
👍